### PR TITLE
feat: Upload images on copy-paste

### PR DESCRIPTION
**Pull request by istarkov**
## Description

Markdown paste with images

Take MD from 
https://github.com/webstudio-is/webstudio/pull/3569#issuecomment-2171213517

Paste, see images are uploaded

## Steps for reproduction

1. click button
2. expect xyz

## Code Review

- [ ] hi @kof, I need you to do
  - conceptual review (architecture, feature-correctness)
  - detailed review (read every line)
  - test it on preview

## Before requesting a review

- [ ] made a self-review
- [ ] added inline comments where things may be not obvious (the "why", not "what")

## Before merging

- [ ] tested locally and on preview environment (preview dev login: 5de6)
- [ ] updated [test cases](https://github.com/webstudio-is/webstudio/blob/main/apps/builder/docs/test-cases.md) document
- [ ] added tests
- [ ] if any new env variables are added, added them to `.env` file

### CODE REVIEW

File: apps/builder/app/shared/copy-paste/asset-upload.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
+import type { WebstudioFragment } from "@webstudio-is/sdk";
+import { builderApi } from "../builder-api";
+
+type StringProp = Extract<
+  WebstudioFragment["props"][number],
+  { type: "string" }
+>;
+
+export const denormalizeSrcProps = async (
```
Comment (kof) [inline apps/builder/app/shared/copy-paste/asset-upload.ts:9]: this could have unit tests, no?

**Review by TrySound — APPROVED** [review]

**Comment by istarkov** [comment]
```md
# HELLO

![1474747aae0ac398812435209b6241e7 webp](https://github.com/webstudio-is/webstudio/assets/5077042/d5798ea3-59e1-4b0d-b741-8e725bafb347)
```

File: apps/builder/app/routes/rest.assets.$name.tsx [hunk 2]
```diff
@@ -22,6 +22,8 @@ export const action = async (
     if (request.method === "POST" && request.body !== null) {
       let body = request.body;
 
+      await new Promise((r) => setTimeout(r, 3000));
```
Comment (kof) [inline apps/builder/app/routes/rest.assets.$name.tsx:25]: What is it doing?

File: apps/builder/app/shared/copy-paste/asset-upload.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    )
+    .map(
+      (prop) =>
+        [prop.id, new URL(prop.value).href] as [propId: string, href: string]
+    );
+
+  return srcProps;
+};
+
+export const denormalizeSrcProps = async (
```
Comment (kof) [inline apps/builder/app/shared/copy-paste/asset-upload.ts:37]: needs unit tests?

Comment (kof) [reply apps/builder/app/shared/copy-paste/asset-upload.ts:37]: needs description what this is doing, hard to tell

**Comment by kof** [comment]
try this

```markdown
# Image Component

![Untitled](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/cb8f4b5b-b3a7-40d7-b19f-f84980dafd60/Untitled.png)

In this article, we will learn how to render optimized, responsive images for a site using the Image Component in Webstudio.
```

File: apps/builder/app/builder/shared/assets/use-assets.tsx [hunk 4]
```diff
@@ -269,65 +278,165 @@ const ToastMessage = ({
...
+      const assetId = fileData.assetId;
+
+      if ($assets.get().has(assetId)) {
+        toast.warn(
+          <ToastMessage
+            title="Info"
+            assetId={assetId}
+            objectURL={fileData.objectURL}
+          >
+            Asset is already exists
```
Comment (kof) [inline apps/builder/app/builder/shared/assets/use-assets.tsx:318]: please check grammar

File: apps/builder/app/builder/shared/assets/use-assets.tsx [hunk 5]
```diff
@@ -269,65 +278,165 @@ const ToastMessage = ({
...
       toast.warn(
         <ToastMessage
-          title="Info"
-          assetId={assetId}
+          title="Warning"
+          assetId={fileData.assetId}
           objectURL={fileData.objectURL}
         >
-          Asset is already exists
+          Asset is already being uploaded
```
Comment (kof) [inline apps/builder/app/builder/shared/assets/use-assets.tsx:403]: I see this message while the upload was already done, I am not sure what is meant here, that the image has been already uploaded? "being" is incorrect in this case, its already been uploaded

**Comment by kof** [comment]
- [ ] info tooltip can't be dismissed by clicking on it and its staying annoyingly long, I think it should be dismissable

**Review by kof — APPROVED** [review]
Approving, but needs fixing

File: apps/builder/app/builder/shared/assets/use-assets.tsx [hunk 6]
```diff
@@ -38,37 +40,57 @@ export const deleteAssets = (assetIds: Asset["id"][]) => {
   });
 };
 
-const setAsset = (asset: Asset) => {
+const safeDeleteAssets = (assetIds: Asset["id"][], projectId: string) => {
+  const currentProjectId = $project.get()?.id;
+
+  if (currentProjectId !== projectId) {
+    toast.error("Project has been changed, files will not be uploaded");
+    // Can cause data corrupiton
```
Comment (kof) [inline apps/builder/app/builder/shared/assets/use-assets.tsx:48]:
```suggestion
    // Can cause data corruption
```

File: apps/builder/app/builder/shared/assets/use-assets.tsx [hunk 7]
```diff
@@ -269,65 +278,165 @@ const ToastMessage = ({
...
-            assetId={assetId}
-            objectURL={fileData.objectURL}
-          >
-            {error}
-          </ToastMessage>
-        );
-      },
-    });
+  const uniqFilesData = [...uniqFilesDataMap.values()];
+  // Now exclude and warn duplicates if exists in assets or uploading files
```
Comment (kof) [inline apps/builder/app/builder/shared/assets/use-assets.tsx:410]: grammar

File: apps/builder/app/routes/rest.assets.$name.tsx (hunk 2 above)
Comment (istarkov) [reply apps/builder/app/routes/rest.assets.$name.tsx:25]: To show how it works with slows upload

**Pull request merged**